File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n=nums.size();
        int pre[n];
        int suf[n];
        int pr=1,su=1;
        for(int i=0;i<n;i++) {
            pr=pr*nums[i];
            su=su*nums[nums.size()-1-i];
            pre[i] = pr;
            suf[nums.size()-1-i]=su;
        }
        vector<int>ans;
        for(int i=0;i<nums.size();i++) {
            if(i==0) {
                ans.push_back(suf[i+1]);
            }else if(i==nums.size()-1) ans.push_back(pre[i-1]);
            else ans.push_back(pre[i-1]*suf[i+1]);
        }
        return ans;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int ans[] = new int[n];
        int pre=1,suf=1;
        for(int i=0;i<n;i++) {
            ans[i]=pre;
            pre*=nums[i];
        }
        for(int i=n-1;i>=0;i--) {
            ans[i]*=suf;
            suf*=nums[i];
        }
        return ans;
    }
}